src/ck-core/charm.h

```diff
@@ -426,7 +426,19 @@ extern void CkArrayExtSend_multi(int aid, int *idx, int ndims, int epIdx, int nu
 
 /*@}*/
 
+typedef CMK_REFNUM_TYPE CkFutureID;
+typedef struct _ckFuture {
+  CkFutureID id;
+  int        pe;
```

Comment:
This does seem pretty trivially convertible to 64bit ID if that's something we want to pursue, which is good. Current 64bit ID formate is [TAG | PE | Element ID], where TAG is a 3bit tag specifying the type of the ID (future in this case).

Reply:
Do other examples come to mind where this type of abstraction would work well? This seemed like a natural place to experiment with it because it's doing multiple reductions -- but putting tests elsewhere would be fine as well. The futures example isn't really doing anything where this sort of callback mechanism makes sense (it's a Fibonacci program) but something that just hammers the futures API could be written as well, I suppose.

Indeed, it maps quite nicely to that formulation; however, do note my response on the other PR regarding its use as a refnum.